### PR TITLE
Shorten implementation of Axes methods that just wrap Axis methods.

### PR DESCRIPTION
**Pull request by anntzer**
... by autogenerating wrappers that forward to the Axis methods.

Admittedly docstring search/replace is a bit ugly, but this has to be
compared with having 4 copies of the docstrings for each method (Axis,
xaxis, yaxis, zaxis) each to be kept in sync.

(As an anecdote, the ability to do this templatization for mplot3d was what originally motivated me to check whether minor ticks could easily be fixed for 3d axes...)

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/matplotlib/axes/_base.py [hunk 1]
```diff
@@ -29,6 +30,54 @@
...
+                 f"contain {k!r} as a substring.")
+            doc = doc.replace(k, v)
+        wrapper.__doc__ = inspect.cleandoc(doc)
+
+    class Placeholder:
+        def __set_name__(self, owner, name):
+            wrapper.__module__ = owner.__module__
+            wrapper.__name__ = name
+            wrapper.__qualname__ = f"{owner.__qualname__}.{name}"
+            setattr(owner, name, wrapper)
```
Comment (tacaswell) [inline lib/matplotlib/axes/_base.py:76]: Does this place holder replace it's self on the wraper on the class as part of class definition?   I assume this is done to make sure the reported names are all correct?

Comment (anntzer) [reply lib/matplotlib/axes/_base.py:76]: Yes.

**Comment by anntzer** [comment]
I pushed a slightly different implementation (directly lifting the "Placeholder" class out) and left some comments explaining why this relies on `__set_name__` and how it works.

File: lib/matplotlib/axes/_base.py [hunk 2]
```diff
@@ -29,6 +30,61 @@
...
+
+    After ::
+
+        get_foo = _axis_method_wrapper("xaxis", "get_bar")
+
+    (in the body of a class) ``get_foo`` is a method that forwards it arguments
+    to the ``get_bar`` method of the ``xaxis`` attribute, and gets its
+    signature and docstring from ``Axis.get_bar``.
+
+    The docstring of ``get_foo`` is built by replacing "this Axis" by "the
```
Comment (timhoffm) [inline lib/matplotlib/axes/_base.py:45]: I'm slightly worried that subsequent edits on the docstring will break this clever replacements. People usually won't know that "this Axis" has a special meaning. They could easily write:
- `this axis`
- ``this `.Axis` ``
- `the axis`

I'm afraid that we can't do much in terms of documenting this for docstring authors or warning upon processing. But let's at least include the first two variants for replacement.

File: lib/matplotlib/axes/_base.py [hunk 3]
```diff
@@ -29,6 +30,61 @@
...
+    {attr_name}" ("the xaxis", "the yaxis") in the wrapped method's docstring;
```
[the remark this answers is not in the document]
Comment (anntzer) [reply lib/matplotlib/axes/_base.py:46]: sure

File: lib/matplotlib/axes/_base.py (hunk 2 above)
Comment (anntzer) [reply lib/matplotlib/axes/_base.py:45]: I'd rather keep things consistent.  Note that there is an assertion check that the correct substring is present, with a hopefully clear error message -- and the check occurs at import time, so we'd easily catch it in CI.

Comment (timhoffm) [reply lib/matplotlib/axes/_base.py:45]: Ok, I did not realize the assert.

**Comment by timhoffm** [comment]
Lot's of warnings in the sphinx build

~~~
/home/circleci/project/lib/matplotlib/patches.py:docstring of matplotlib.patches.ArrowStyle.BarAB:1: WARNING: py:class reference target not found: matplotlib.patches.ArrowStyle._Bracket
/home/circleci/project/lib/matplotlib/patches.py:docstring of matplotlib.patches.ArrowStyle.BracketA:1: WARNING: py:class reference target not found: matplotlib.patches.ArrowStyle._Bracket
/home/circleci/project/lib/matplotlib/patches.py:docstring of matplotlib.patches.ArrowStyle.BracketAB:1: WARNING: py:class reference target not found: matplotlib.patches.ArrowStyle._Bracket
/home/circleci/project/lib/matplotlib/patches.py:docstring of matplotlib.patches.ArrowStyle.BracketB:1: WARNING: py:class reference target not found: matplotlib.patches.ArrowStyle._Bracket
/home/circleci/project/lib/matplotlib/patches.py:docstring of matplotlib.patches.ArrowStyle.Curve:1: WARNING: py:class reference target not found: matplotlib.patches.ArrowStyle._Curve
/home/circleci/project/lib/matplotlib/patches.py:docstring of matplotlib.patches.ArrowStyle.CurveA:1: WARNING: py:class reference target not found: matplotlib.patches.ArrowStyle._Curve
/home/circleci/project/lib/matplotlib/patches.py:docstring of matplotlib.patches.ArrowStyle.CurveAB:1: WARNING: py:class reference target not found: matplotlib.patches.ArrowStyle._Curve
/home/circleci/project/lib/matplotlib/patches.py:docstring of matplotlib.patches.ArrowStyle.CurveB:1: WARNING: py:class reference target not found: matplotlib.patches.ArrowStyle._Curve
/home/circleci/project/lib/matplotlib/patches.py:docstring of matplotlib.patches.ArrowStyle.CurveFilledA:1: WARNING: py:class reference target not found: matplotlib.patches.ArrowStyle._Curve
/home/circleci/project/lib/matplotlib/patches.py:docstring of matplotlib.patches.ArrowStyle.CurveFilledAB:1: WARNING: py:class reference target not found: matplotlib.patches.ArrowStyle._Curve
/home/circleci/project/lib/matplotlib/patches.py:docstring of matplotlib.patches.ArrowStyle.CurveFilledB:1: WARNING: py:class reference target not found: matplotlib.patches.ArrowStyle._Curve
/home/circleci/project/lib/matplotlib/patches.py:docstring of matplotlib.patches.ArrowStyle.Fancy:1: WARNING: py:class reference target not found: matplotlib.patches.ArrowStyle._Base
/home/circleci/project/lib/matplotlib/patches.py:docstring of matplotlib.patches.ArrowStyle.Simple:1: WARNING: py:class reference target not found: matplotlib.patches.ArrowStyle._Base
/home/circleci/project/lib/matplotlib/patches.py:docstring of matplotlib.patches.ArrowStyle.Wedge:1: WARNING: py:class reference target not found: matplotlib.patches.ArrowStyle._Base
/home/circleci/project/lib/matplotlib/patches.py:docstring of matplotlib.patches.BoxStyle.Circle:1: WARNING: py:class reference target not found: matplotlib.patches.BoxStyle._Base
/home/circleci/project/lib/matplotlib/patches.py:docstring of matplotlib.patches.BoxStyle.DArrow:1: WARNING: py:class reference target not found: matplotlib.patches.BoxStyle._Base
/home/circleci/project/lib/matplotlib/patches.py:docstring of matplotlib.patches.BoxStyle.LArrow:1: WARNING: py:class reference target not found: matplotlib.patches.BoxStyle._Base
/home/circleci/project/lib/matplotlib/patches.py:docstring of matplotlib.patches.BoxStyle.Round:1: WARNING: py:class reference target not found: matplotlib.patches.BoxStyle._Base
/home/circleci/project/lib/matplotlib/patches.py:docstring of matplotlib.patches.BoxStyle.Round4:1: WARNING: py:class reference target not found: matplotlib.patches.BoxStyle._Base
/home/circleci/project/lib/matplotlib/patches.py:docstring of matplotlib.patches.BoxStyle.Sawtooth:1: WARNING: py:class reference target not found: matplotlib.patches.BoxStyle._Base
/home/circleci/project/lib/matplotlib/patches.py:docstring of matplotlib.patches.BoxStyle.Square:1: WARNING: py:class reference target not found: matplotlib.patches.BoxStyle._Base
/home/circleci/project/lib/matplotlib/patches.py:docstring of matplotlib.patches.ConnectionStyle.Angle:1: WARNING: py:class reference target not found: matplotlib.patches.ConnectionStyle._Base
/home/circleci/project/lib/matplotlib/patches.py:docstring of matplotlib.patches.ConnectionStyle.Angle3:1: WARNING: py:class reference target not found: matplotlib.patches.ConnectionStyle._Base
/home/circleci/project/lib/matplotlib/patches.py:docstring of matplotlib.patches.ConnectionStyle.Arc:1: WARNING: py:class reference target not found: matplotlib.patches.ConnectionStyle._Base
/home/circleci/project/lib/matplotlib/patches.py:docstring of matplotlib.patches.ConnectionStyle.Arc3:1: WARNING: py:class reference target not found: matplotlib.patches.ConnectionStyle._Base
/home/circleci/project/lib/matplotlib/patches.py:docstring of matplotlib.patches.ConnectionStyle.Bar:1: WARNING: py:class reference target not found: matplotlib.patches.ConnectionStyle._Base
/home/circleci/project/lib/mpl_toolkits/axisartist/axisline_style.py:docstring of mpl_toolkits.axisartist.axisline_style.AxislineStyle.SimpleArrow:1: WARNING: py:class reference target not found: mpl_toolkits.axisartist.axisline_style.AxislineStyle._Base
/home/circleci/project/lib/mpl_toolkits/axisartist/axislines.py:docstring of mpl_toolkits.axisartist.axislines.AxisArtistHelper.Fixed:1: WARNING: py:class reference target not found: mpl_toolkits.axisartist.axislines.AxisArtistHelper._Base
/home/circleci/project/lib/mpl_toolkits/axisartist/axislines.py:docstring of mpl_toolkits.axisartist.axislines.AxisArtistHelper.Floating:1: WARNING: py:class reference target not found: mpl_toolkits.axisartist.axislines.AxisArtistHelper._Base
~~~

I think they are unrelated, but not sure about it.

**Comment by anntzer** [comment]
I'd believe it's something to do with sphinx 3.0, but I'm fine waiting for that to be handled first (e.g. by temporarily pinning to <3.0) before merging this.

**Pull request merged**